azure-core/azure/core/pipeline/transport/requests_basic.py

```diff
@@ -0,0 +1,205 @@
+    HttpResponse,
+    _HttpResponseBase
+)
+
+from azure.core.configuration import Configuration
+from azure.core.exceptions import (
+    ServiceRequestError,
+    ConnectionError,
+    ConnectionReadError,
+    ReadTimeoutError,
+    raise_with_traceback
+)
+
+# Matching requests, because why not?
+CONTENT_CHUNK_SIZE = 10 * 1024
+
+class RequestsContext(object):
+    def __init__(self, session, transport):
+        self.session = session
+        self.transport = transport
+
+
+class _RequestsTransportResponseBase(_HttpResponseBase):
+
+    def __init__(self, request, requests_response):
+        super(_RequestsTransportResponseBase, self).__init__(request, requests_response)
+        self.status_code = requests_response.status_code
+        self.headers = requests_response.headers
+        self.reason = requests_response.reason
+
+    def body(self):
+        return self.internal_response.content
+
+    def text(self, encoding=None):
+        if encoding:
+            self.internal_response.encoding = encoding
+        return self.internal_response.text
+
+    def raise_for_status(self):
+        self.internal_response.raise_for_status()
+
+
+class RequestsTransportResponse(_RequestsTransportResponseBase, HttpResponse):
+
+    def stream_download(self, chunk_size=None, callback=None):
+        # type: (Optional[int], Optional[Callable]) -> Iterator[bytes]
+        """Generator for streaming request body data.
+
+        :param callback: Custom callback for monitoring progress.
+        :param int chunk_size:
+        """
+        chunk_size = chunk_size or CONTENT_CHUNK_SIZE
+        with contextlib.closing(self.internal_response) as response:
+            # https://github.com/PyCQA/pylint/issues/1437
+            for chunk in response.iter_content(chunk_size):  # pylint: disable=no-member
+                if not chunk:
+                    break
+                if callback and callable(callback):
+                    callback(chunk, response=response)
+                yield chunk
+
+
+class RequestsTransport(HttpTransport):
+    """Implements a basic requests HTTP sender.
+
+    Since requests team recommends to use one session per requests, you should
+    not consider this class as thread-safe, since it will use one Session
+    per instance.
+
+    In this simple implementation:
+    - You provide the configured session if you want to, or a basic session is created.
+    - All kwargs received by "send" are sent to session.request directly
+    """
+
+    _protocols = ['http://', 'https://']
+
+    def __init__(self, configuration=None, session=None):
+        # type: (Optional[requests.Session]) -> None
+        self._session_mapping = threading.local()
+        self.config = configuration or Configuration()
+        self.session = session or requests.Session()
+
+    def __enter__(self):
+        # type: () -> RequestsTransport
+        return self
+
+    def __exit__(self, *exc_details):  # pylint: disable=arguments-differ
```

Comment:
The combination of being able to pass in a session - or owning a session if one is not passed in - is somewhat confusing to me. It is not clear exactly what the expected lifetime management of the session is.

Reply:
I'm wondering if we need to be able to pass in a session at all.... I can't think of a scenario where this would be needed?

Reply:
Sometimespeople wants to set proxy and all complicated stuff we don't wrap, hence the session.
If you give it to me, I own it (that's my semantic suggestion)

Reply:
But we already expose proxy config... I'm curious what other options on a session would people need to configure? Maybe I'll take a closer look at the object to see what other properties it has...
If people really want to do super convoluted things - like patch the session object - I think they could make a case for supplying their own custom "Transport" flavor.